provider/vsphere/internal/vsphereclient/client.go

```diff
@@ -259,16 +279,13 @@ func (c *Client) ResourcePools(ctx context.Context, path string) ([]*object.Reso
 }
 
 // EnsureVMFolder creates the a VM folder with the given path if it doesn't
-// already exist.
-func (c *Client) EnsureVMFolder(ctx context.Context, folderPath string) (*object.Folder, error) {
+// already exists.
+func (c *Client) EnsureVMFolder(ctx context.Context, folderPath string, parentFolderStr string) (*object.Folder, error) {
+
 	finder, datacenter, err := c.finder(ctx)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	folders, err := datacenter.Folders(ctx)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
 
 	createFolder := func(parent *object.Folder, name string) (*object.Folder, error) {
 		folder, err := parent.CreateFolder(ctx, name)
```

Comment:
If you had EnsureVMFolder just take one path then this whould change to use your findFolder method, so that rather than trying to create the folder and handling the DuplicateName error it just checks for the path and only tries to create it if it doesn't exist.

Reply:
@babbageclunk there are two ways we can work this out. Either I expose FindFolder and use it on the config.go's VMFolder method. That means each time we call this function, go_routine will sleep waiting for govc. I personally don't like that.

I'd rather keep findFolder private within vsphereclient and use it only once, when we define EnsureVMFolder (i.e. create the folder). Is that ok? I will swap strings as you've suggested anyway.